ocsp/responder/redis/checked_redis_source.go

```diff
@@ -0,0 +1,137 @@
+package redis
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/go-gorp/gorp/v3"
+	"github.com/letsencrypt/boulder/core"
+	"github.com/letsencrypt/boulder/db"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/ocsp/responder"
+	"github.com/letsencrypt/boulder/sa"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/ocsp"
+)
+
+// dbSelector is a limited subset of the db.WrappedMap interface to allow for
+// easier mocking of mysql operations in tests.
+type dbSelector interface {
+	SelectOne(holder interface{}, query string, args ...interface{}) error
+	WithContext(ctx context.Context) gorp.SqlExecutor
+}
+
+// rocspSourceInterface expands on responder.Source by adding a private signAndSave method.
+// This allows checkedRedisSource to trigger a live signing if the DB disagrees with Redis.
+type rocspSourceInterface interface {
+	Response(ctx context.Context, req *ocsp.Request) (*responder.Response, error)
+	signAndSave(ctx context.Context, req *ocsp.Request, cause string) (*responder.Response, error)
+}
+
+type checkedRedisSource struct {
+	base    rocspSourceInterface
+	dbMap   dbSelector
+	counter *prometheus.CounterVec
+	log     blog.Logger
+}
+
+func NewCheckedRedisSource(base *redisSource, dbMap dbSelector, stats prometheus.Registerer, log blog.Logger) *checkedRedisSource {
+	if base == nil {
+		return nil
+	}
+
+	return newCheckedRedisSource(base, dbMap, stats, log)
+}
+
+// Internal-only method that takes a private interface as a parameter. We call this from tests and
+// from NewCheckedRedisSource.
+func newCheckedRedisSource(base rocspSourceInterface, dbMap dbSelector, stats prometheus.Registerer, log blog.Logger) *checkedRedisSource {
+	counter := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "checked_rocsp_responses",
+		Help: "Count of OCSP requests/responses from checkedRedisSource, by result",
+	}, []string{"result"})
+	stats.MustRegister(counter)
+
+	return &checkedRedisSource{
+		base:    base,
+		dbMap:   dbMap,
+		counter: counter,
+		log:     log,
+	}
+}
+
+var ocspCodeToStatus = map[int]core.OCSPStatus{
+	ocsp.Good:    core.OCSPStatusGood,
+	ocsp.Revoked: core.OCSPStatusRevoked,
+}
+
+// Response implements the responder.Source interface. It looks up the requested OCSP
+// response in the redis cluster and looks up the corresponding status in the DB. If
+// the status disagrees with what redis says, it signs a fresh response and serves it.
+func (src *checkedRedisSource) Response(ctx context.Context, req *ocsp.Request) (*responder.Response, error) {
+	serialString := core.SerialToString(req.SerialNumber)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	var dbStatus core.CertificateStatus
+	var redisResult *responder.Response
+	var redisErr, dbErr error
+	go func() {
+		defer wg.Done()
+		// TODO: Make a more efficient query here that looks only at status and revokedAt.
```

Comment:
Attach to and issue.

```suggestion
		// TODO: Make a more efficient query here that looks only at status and revokedAt.
```